feat: add `poo/abstraction/en` (#28)
readthedocs-fr/notions: poo/abstraction/en/Abstraction.md

# Abstraction

> The **abstraction** idea, is chopper a class in parts, to be able to share (via a parameter passed in method for example) **an only** of this part rather the object “in full”.

Reviewer: Not sure about the `chopper` translation. Maybe `divide` or `cut out` are better translations.
Also I think `only one` is a better translation than `an only`.

Author: Tanks, the changes have been made!

To understand the utility of this principle and how apply, take for example a class Human, since this is a thing that we know better.

```java

class Human {

    int air;
    int satiety;
    float energy;

    void breathe() {
        air += 10;
    }

    void eat() {
        satiety += 5;
    }

    void move() {
        System.out.println("Un humain souhaite se déplacer !");
    }

    void consumeEnergy(float value) {
        this.energie -= value;
    }
}
```

> Now, imagine that we have a class Hiking which allows organise a walk enter Humans. A thing of style:

Reviewer: I think this part is incorrect:
> a class Hiking which allows organise a walk enter Humans

I would have translated that as ```a class `Hiking` that allows us to organise/organize a walk[...]```
Also the text is not formatted as the original. `Hiking` is in code block and Human is not written as a class in the French version.
I need help to translate `une marche entre humains` because I'm really not certain with the current translation `a walk enter Humans`.

Author: The text formatting has been checked and corrected
`a class Hiking which allows organise a walk` has been replace to `a class Hiking that allows us to organise a walk`

Reviewer: Thanks

```java
class Hiking {

    List<Human> participants;

    public Hiking() { this.participants = new ArrayList<>(); }

    void addParticipant(Human participant) { this.participants.add(participant); }

    void moveAll() {
        for(Human human : participants) {
            human.move();
            human.consumeEnergy(3);
        }
    }
}
```

> So far so good, we have that we want. This code works perfectly, except, in reality, he contain two major problems.

Reviewer: Bad pronoun and conjugation:
- `This code [...] he contain [...]` -> `This code [...] it contains [...]`
```suggestion
So far so good, we have that we want. This code works perfectly, except, in reality, it contains two major problems.

```

Author: Corrected, tanks!

1. **We restraint the hiking of the objects of type Human**, whereas anything can walk and exhausted could take a hiking. It’s not really the kind to forbid to elephants and to tractors to participate :(

> 2. We ask a human in full but us all we need, this is his legs (hum). This class hiking, however currently access to the human respiratory system (via `breathe()` ) and to this stomach (via `eat()` ) ! **So, we give too much power to this class Hiking, since she is able to do things who shouldn’t be her accessible.**

Reviewer: - Maybe `a whole human` is a better translation than `a human in full`.
- Also, reformulate `but us all we need, this is his legs` into `but all we need is his legs`. Maybe use `their` in replacement for `his` because we can't really use a gendered pronoun for an unknown human.
- Also `to this stomach` (`à cette estomac`) is incorrect. You meant to say `to his/her/their stomach`.
- Finally, a class is not a living person, so its pronoun is `it` (`she is able to` -> `it is able to`)
```suggestion
2. We ask a whole human but all we need is their legs (hum). This class hiking, however currently access to the human respiratory system (via `breathe()` ) and to their stomach (via `eat()` ) ! **So, we give too much power to this class Hiking, since it is able to do things who shouldn’t be accessible to it.**

```

Author: I applied the modifications,

>  a human in full
has been replaced to
>  a whole human
--
> but us all we need, this is his legs
by
> but all we need is his legs
--
Oups... a `t` slipped 😅
> to this stomach
has been replaced by
> to his stomach
--
> she is able to
has been replaced by
> it is able to
Thank for this details!

Reviewer: Cool! _BTW, you should skip a line after quote to exit the quote block in Github_

> So, how to contour this two problems? 

Reviewer: Typo:
`contour` -> `get around`
```suggestion
So, how to get around this two problems? 

```

Author: Corrected! Tanks!

```java
class Human implements Walking {

    int air;
    int satiety;
    float energy;

    void breathe() {
        air += 10;
    }

    void eat() {
        satiety += 5;
    }

    @Override
    void move() {
        System.out.println("Un humain souhaite se déplacer !");
    }

    @Override
    void consumeEnergy(float value) {
        this.energy -= value;
    }
}
```
But, what to change? Now, instead of making a human list in our Hiking, we will do a *Walking list* (it’s more politically correct). And, there is almost nothing to change, it’s a good sign:

```java
class Hiking {

    List<Walking> participants;

    public Hiking() { this.participants = new ArrayList<>(); }

    void addParticipant(Walking participant) { this.participants.add(participant); }

    void moveAll() {
        for(Walking walking: participants) {
            walking.move();
            walking.consumeEnergy(3);
        }
    }
}
```

> And, henceforth, not only our Hiking class hasn’t access to the Human methods `breathe` and `eat`, since, nowhere, we said to our work with humans (no, but think, if a hiking can make you swallow peas, it would be creepy, still), but more, she accepts the elephants and the tractors! Provided that implements `Walking` them too… and, who know how to out of print a tractor.

Reviewer: Bad pronoun for the class (`she` -> `it)
Fix `that` -> `they`
`
```suggestion
And, henceforth, not only our Hiking class hasn’t access to the Human methods `breathe` and `eat`, since, nowhere, we said to our work with humans (no, but think, if a hiking can make you swallow peas, it would be creepy, still), but more, it accepts the elephants and the tractors! Provided they implement `Walking` them too… and, who know how to out of print a tractor.

```

Author: Corrected!

Thank you for your time and correction!
How was my translation? Very bad or can it pass?
Do you think that if I translate another document it will be beneficial or a waste of time, especially for you? :)

Reviewer: It's a good translation overall. Even in French documents, we have several comments to make. 
Your translation is really correct. And even with the typo, it will always be beneficial.

Author: Ok! Very thank, have a good evening!

# Abstraction

**Abstraction** is the idea of dividing a class into several parts, to be able to share (via a parameter passed in a method for example) **only one** of these parts, rather than the object entirely. To understand the utility of this principle and how to apply it, take for example a class `Human` since this is something that we know well.

```java

class Human {

    int air;
    int satiety;
    float energy;

    void breathe() {
        air += 10;
    }

    void eat() {
        satiety += 5;
    }

    void move() {

>         System.out.println("Un humain souhaite se déplacer !");

Reviewer: ```suggestion
        System.out.println("A human wants to move!");

```

    }

    void consumeEnergy(float value) {

>         this.energie -= value;

Reviewer 2: ```suggestion
        this.energy -= value;
```

Author: Fixed

    }
}
```

Now, imagine that we have a class `Hiking` that allows us to organise a walk. Something like:

```java

class Hiking {

    List<Human> participants;

    public Hiking() { this.participants = new ArrayList<>(); }

    void addParticipant(Human participant) { this.participants.add(participant); }

    void moveAll() {
        for(Human human : participants) {
            human.move();
            human.consumeEnergy(3);
        }
    }
}
```

So far so good, we have what we want. This code works perfectly well, except that, in reality, it contains two major problems.

> 1. **We restraint the hiking of the objects of type Human**, whereas anything that can **walk** and **be exhausted** could take a hike. It’s not really the kind to forbid to elephants and to tractors to participate :(

Reviewer 2: Why is there a `(` here?

Author: This is a smiley, the same of the fr version
> 1. On restreint la randonnée aux objets de type Human, alors que n'importe quoi qui peut marcher et s'épuiser pourrait faire une randonnée. C'est vraiment pas gentil d'interdire aux éléphants et aux tracteurs de participer :(

Reviewer 2: Oh okay 😂. On my phone I just saw a `(`

Author: No problem! 😄

> 2. We ask a whole human but us all we need is their legs (hum). This class hiking, however currently access to the human respiratory system (via `breathe()` ) and to his stomach (via `eat()` ) ! **So, we give too much power to this class Hiking, since it is able to do things who shouldn’t be accessible to it**. <br>

Reviewer 3: ```suggestion
2. We ask for a whole human but all we need is their legs (hum). This hiking class, however currently has access to the human breathing system (via `breathe()` ) and to his stomach (via `eat()` )! **So, we give too much power to this class, since it is able to do things that it shouldn’t be able to do**. <br>

```

> So, how to get around this two problems? 

Reviewer: Typo in the suggestion

> So, how to get around this two problems? 

Reviewer 3: Fixed, ty

> The **interfaces** come to our rescue (interfaces are great). Let's make a `Walking` interface, which allows to describe something that moves and gets tired:

Reviewer 3: ```suggestion
**Interfaces** come to our rescue (interfaces are great). Let's make a `Walking` interface, which allows describing something that moves and gets tired:

```

```java
interface Walking {
    void move();
    void consumeEnergy(float value);
}
```

> So, now, our class Human will be able to **implement** this super interface, and in reality we add `implements Walking` and put a little `@Override` on their methods `move` and `consumeEnergy` and the turn is played, since this is a behaviour that we had already implemented beforehand, just without the interface.

Reviewer 3: ```suggestion
So now, our class `Human` will be able to **implement** this awesome interface, and all that's left to do is to add `implements Walking` and put a little `@Override` on our `move` and `consumeEnergy` methods, and that's it, since this is a behavior that we had already implemented beforehand, only without the interface.

```
I kinda rewrote this section, since I couldn't quite understand some parts that you translated. Also, since it is more common to write in American style, I changed `behaviour` to `behavior`

```java
class Human implements Walking {

    int air;
    int satiety;
    float energy;

    void breathe() {
        air += 10;
    }

    void eat() {
        satiety += 5;
    }

    @Override
    void move() {

>         System.out.println("Un humain souhaite se déplacer !");

Reviewer: ```suggestion
        System.out.println("A human wants to move!");

```

    }

    @Override
    void consumeEnergy(float value) {
        this.energy -= value;
    }
}
```

> But, what to change? Now, instead of making a human list in our Hiking, we will do a **Walking list** (it’s more politically correct). And, there is almost nothing to change, it’s a good sign:

Reviewer 3: ```suggestion
But, what has changed? Well, now instead of making a Human list in our Hiking class, we will do a **Walking list** (it’s more politically correct). And, there is almost nothing to change, which is a good sign:

```

Reviewer: I think you misunderstood the sentence. Maybe I'm wrong but I can't check for the french version now

```java

class Hiking {

    List<Walking> participants;

    public Hiking() { this.participants = new ArrayList<>(); }

    void addParticipant(Walking participant) { this.participants.add(participant); }

    void moveAll() {
        for(Walking walking: participants) {
            walking.move();
            walking.consumeEnergy(3);
        }
    }
}
```

> And, henceforth, not only our Hiking class hasn’t access to the Human methods `breathe` and `eat`, since, nowhere, we said to our work with humans (no, but think, if a hiking can make you swallow peas, it would be creepy, still), but more, it accepts the elephants and the tractors !

Reviewer 3: ```suggestion
And, henceforth, not only doesn’t our Hiking class have access to the Human methods `breathe` and `eat`, since, we never specified that it's humans we're working with (think about it, if hiking could make you swallow peas, it would be creepy, right?), but also it accepts the elephants and the tractors!

```
`not only ... but also` has a pretty nasty syntax, so I corrected the few mistakes you made here.

> Provided they implement `Walking` them too… and, who know how to out of print a tractor.

Reviewer 3: ```suggestion
But that would assuming that they implement `Walking` too… and, who knows how to make a tractor tired.

```

## Conclusion

> Make an abstraction to our **Human** class we allowed to do two very important opposite things:

Reviewer 3: ```suggestion
By making an abstraction to our **Human** class, we allowed two very important yet opposite things:

```

> - **An access restriction**, since `Hiking` hasn’t access to some functions to the `Human` class

Reviewer 3: ```suggestion
- **An access restriction**, since `Hiking` doesn’t have access to the functions that are specific to the `Human` class

```

> - **An ease of access**, since `Hiking` accepts henceforth, every object type implementing `Walking`, instead of being limited to humans.

Reviewer 3: ```suggestion
- **An ease of access**, since `Hiking` accepts henceforth every object whose type implements `Walking`, instead of being limited to humans.

```

> For more information on this principle [SOLID] (<https://fr.wikipedia.org/wiki/SOLID_(informatique)>) named **Dependency Inversion** (who is mainly based on abstraction), I invite you to read [Wikipedia](https://en.wikipedia.org/wiki/Dependency_inversion_principle) on this subject.

Reviewer 3: ```suggestion
For more information on the [SOLID](<https://en.wikipedia.org/wiki/SOLID_(object-oriented_design))>) principle  called **Dependency Inversion** (which is mainly based on abstraction), I invite you to read the [Wikipedia](https://en.wikipedia.org/wiki/Dependency_inversion_principle) article related to this topic.

```

> For more information on the [SOLID] (<https://en.wikipedia.org/wiki/SOLID_(object-oriented_design)>) principle called **Dependency Inversion** (witch is mainly based on abstraction), I invite you to read the [Wikipedia](https://en.wikipedia.org/wiki/Dependency_inversion_principle) article related to this topic.

Reviewer 2: ```suggestion
​For more informations on the [SOLID] (<https://en.wikipedia.org/wiki/SOLID_(object-oriented_design)>) principle called **Dependency Inversion** (witch is mainly based on abstraction), I invite you to read the [Wikipedia](https://en.wikipedia.org/wiki/Dependency_inversion_principle) article related to this topic.
```
Maybe it's better with an `s`?

Author: On the original version, information as writte without `s`
> Pour plus d'information sur le principe...

But, it's true, with a `s` is better 😁,
Fixed, thanks!

Reviewer: Information is an uncountable noun iirc. You don't need this `s`.

Author: So, with a `s` or without?

Reviewer: without

Author: It's good!

Reviewer 2: My bad 😅

# Abstraction

> **Abstraction** is idea of dividing a class into several parts, to be able to share (via a parameter passed in a method for example) **only one** of these parts, rather than the object entirely. To understand the utility of this principle and how to apply it, take for example a class `Human` since this is something that we know well.

Reviewer 3: ```suggestion
**Abstraction** is the idea of dividing a class into several parts, to be able to share (via a parameter passed in a method for example) **only one** of these parts, rather than the object entirely. To understand the utility of this principle and how to apply it, take for example a class `Human` since this is something that we know well.

```

}
```

> So now, our class `Human` will be able to **implement** this awesome interface, and all that's to di is to add `implements Walking` and put a little `@Override` on our `move` and `consumeEnergy` methos, and that's it, since this is a behavior that we had already implemented beforehand, only without the interface.

Reviewer 3: ```suggestion
So now, our class `Human` will be able to **implement** this awesome interface, and all that's left to do is to add `implements Walking` and put a little `@Override` on our `move` and `consumeEnergy` methods, and that's it, since this is a behavior that we had already implemented beforehand, only without the interface.

```

}
```

> And, henceforth, not only doesn't our Hiking class have access to the Human methods `breathe` and `eat`, since, we never specified that it's humans we're working with (think about it, if hiking could make you swallow peas, it would be creepy, right?), but also, it accepts the elephants and the tractors!

Reviewer 3: ```suggestion
And, henceforth, not only doesn't our Hiking class have access to the Human methods `breathe` and `eat`, since, we never specified that it's humans we're working with (think about it, if hiking could make you swallow peas, it would be creepy, right?), but also it accepts the elephants and the tractors!

```
no coma, classic trap 😄

But that would assuming that they implement `Walking` too… and, who knows how to make a tractor tired.

## Conclusion

By making an abstraction to our **Human** class, we allowed two very important yet opposite things:

- **An access restriction**, since `Hiking` doesn’t have access to the functions that are specific to the `Human` class
- **An ease of access**, since `Hiking` accepts henceforth every object whose type implements `Walking`, instead of being limited to humans.

> For more information on the [SOLID] (<https://en.wikipedia.org/wiki/SOLID_(object-oriented_design)>) principle called **Dependency Inversion** (witch is mainly based on abstraction), I invite you to read the [Wikipedia](https://en.wikipedia.org/wiki/Dependency_inversion_principle) article related to this topic.

Reviewer 3: ```suggestion
For more information on the [SOLID] (<https://en.wikipedia.org/wiki/SOLID_(object-oriented_design)>) principle called **Dependency Inversion** (which is mainly based on abstraction), I invite you to read the [Wikipedia](https://en.wikipedia.org/wiki/Dependency_inversion_principle) article related to this topic.

```
